### PR TITLE
[Merged by Bors] - feat(Polynomial): Equivalence between monic polynomials of degree n and `degreeLT n`

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---
- [x] depends on: #16031 
- [x] depends on: #16030
- [x] depends on: #16019
- [x] depends on: #16035 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Polynomial/Basic.lean [hunk 1]
```diff
@@ -173,6 +173,34 @@ theorem degreeLT_succ_eq_degreeLE {n : ℕ} : degreeLT R (n + 1) = degreeLE R n
   · rw [mem_degreeLT, mem_degreeLE, ← natDegree_lt_iff_degree_lt (by rwa [ne_eq]),
       ← natDegree_le_iff_degree_le, Nat.lt_succ]
 
+/-- The equivalence between monic polynomials of degree `n` and polynomials of degree less than
+`n`, formed by adding a term `X ^ n`. -/
+def monicEquivDegreeLT [Nontrivial R] (n : ℕ) :
+    { p : R[X] // p.Monic ∧ p.natDegree = n } ≃ degreeLT R n :=
+  { toFun := fun p => ⟨p.1.eraseLead, by
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Polynomial/Basic.lean:180]: I would just advise you use `where` notation
```suggestion
    { p : R[X] // p.Monic ∧ p.natDegree = n } ≃ degreeLT R n where
  toFun p := ⟨p.1.eraseLead, by
```

**Review by YaelDillies — APPROVED** [review]
Looks hardly golfable

maintainer merge

File: Mathlib/RingTheory/Polynomial/Basic.lean [hunk 2]
```diff
@@ -173,6 +173,34 @@ theorem degreeLT_succ_eq_degreeLE {n : ℕ} : degreeLT R (n + 1) = degreeLE R n
...
+        simp only
+        conv_rhs => rw [← eraseLead_add_C_mul_X_pow p]
+        simp [Monic.def.1 hp, add_comm],
+      right_inv := by
+        rintro ⟨p, hp⟩
+        ext1
+        simp only
+        rw [eraseLead_add_of_degree_lt_left]
+        · simp
+        · simp [mem_degreeLT.1 hp] }
```
Comment (ChrisHughes24) [inline Mathlib/RingTheory/Polynomial/Basic.lean:202]:
```suggestion
        · simp [mem_degreeLT.1 hp]
```

File: Mathlib/RingTheory/Polynomial/Basic.lean [hunk 3]
```diff
@@ -173,6 +173,34 @@ theorem degreeLT_succ_eq_degreeLE {n : ℕ} : degreeLT R (n + 1) = degreeLE R n
...
+/-- The equivalence between monic polynomials of degree `n` and polynomials of degree less than
+`n`, formed by adding a term `X ^ n`. -/
+def monicEquivDegreeLT [Nontrivial R] (n : ℕ) :
+    { p : R[X] // p.Monic ∧ p.natDegree = n } ≃ degreeLT R n where
+  toFun p := ⟨p.1.eraseLead, by
+        rcases p with ⟨p, hp, rfl⟩
+        simp only [mem_degreeLT]
+        refine lt_of_lt_of_le ?_ degree_le_natDegree
+        exact degree_eraseLead_lt (ne_zero_of_ne_zero_of_monic one_ne_zero hp)⟩,
+      invFun := fun p =>
```
Comment (jcommelin) [inline Mathlib/RingTheory/Polynomial/Basic.lean:185]: Can you please fix the indentation here and below?

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**